common/src/app/common/geom/shapes/flex_layout/lines.cljc

```diff
@@ -0,0 +1,312 @@
+                child-max-width  (ctl/child-max-width child)
+                child-max-height (ctl/child-max-height child)
+
+                [child-margin-top child-margin-right child-margin-bottom child-margin-left]
+                (ctl/child-margins child)
+
+                child-margin-width (+ child-margin-left child-margin-right)
+                child-margin-height (+ child-margin-top child-margin-bottom)
+
+                fill-width?  (ctl/fill-width? child)
+                fill-height? (ctl/fill-height? child)
+
+                ;; We need this info later to calculate the child resizes when fill
+                child-data {:id (:id child)
+                            :child-min-width (if fill-width? child-min-width child-width)
+                            :child-min-height (if fill-height? child-min-height child-height)
+                            :child-max-width (if fill-width? child-max-width child-width)
+                            :child-max-height (if fill-height? child-max-height child-height)}
+
+                next-min-width   (+ child-margin-width (if fill-width? child-min-width child-width))
+                next-min-height  (+ child-margin-height (if fill-height? child-min-height child-height))
+                next-max-width   (+ child-margin-width (if fill-width? child-max-width child-width))
+                next-max-height  (+ child-margin-height (if fill-height? child-max-height child-height))
+
+                next-line-min-width  (+ line-min-width  next-min-width  (* layout-gap-row num-children))
+                next-line-min-height (+ line-min-height next-min-height (* layout-gap-col num-children))]
+            (if (and (some? line-data)
+                     (or (not wrap?)
+                         (and row? (<= next-line-min-width layout-width))
+                         (and col? (<= next-line-min-height layout-height))))
+
+              [{:line-min-width  (if row? (+ line-min-width next-min-width) (max line-min-width next-min-width))
+                :line-max-width  (if row? (+ line-max-width next-max-width) (max line-max-width next-max-width))
+                :line-min-height (if col? (+ line-min-height next-min-height) (max line-min-height next-min-height))
+                :line-max-height (if col? (+ line-max-height next-max-height) (max line-max-height next-max-height))
+                :num-children    (inc num-children)
+                :children-data   (conjv children-data child-data)}
+               result]
+
+              [{:line-min-width  next-min-width
+                :line-min-height next-min-height
+                :line-max-width  next-max-width
+                :line-max-height next-max-height
+                :num-children    1
+                :children-data   [child-data]}
+               (cond-> result (some? line-data) (conj line-data))])))
+
+        [line-data layout-lines] (reduce calculate-line-data [nil []] children)]
```

Comment:
The same here, if think `loop` will be better on performance (I don't really know if this is on hot path or not, so I just suggest the change)

Reply:
ok